packages/components/src/composite/legacy/index.tsx

```diff
@@ -0,0 +1,177 @@
+/**
+ * Composite is a component that may contain navigable items represented by
+ * CompositeItem. It's inspired by the WAI-ARIA Composite Role and implements
+ * all the keyboard navigation mechanisms to ensure that there's only one
+ * tab stop for the whole Composite element. This means that it can behave as
+ * a roving tabindex or aria-activedescendant container.
+ *
+ * @see https://ariakit.org/components/composite
+ */
+
+/**
+ * WordPress dependencies
+ */
+import {
+	forwardRef,
+	useId as useGeneratedId,
+	useRef,
+	useState,
+} from '@wordpress/element';
+import deprecated from '@wordpress/deprecated';
+
+/**
+ * Internal dependencies
+ */
+import * as Current from '../current';
+
+type Component = ( ...any: any[] ) => any;
+
+type CompositeStore = ReturnType< typeof Current.useCompositeStore >;
+type CompositeStoreState = { store: CompositeStore };
+type CompositeStoreProps = NonNullable<
+	Parameters< typeof Current.useCompositeStore >[ 0 ]
+>;
+
+type Orientation = 'horizontal' | 'vertical';
+
+export interface InitialState {
+	baseId?: string;
+	unstable_virtual?: boolean;
+	rtl?: boolean;
+	orientation?: Orientation;
+	currentId?: string;
+	loop?: boolean | Orientation;
+	wrap?: boolean | Orientation;
+	shift?: boolean;
+}
+
+type BaseIdState = Required< Pick< InitialState, 'baseId' > >;
+type BaseId = BaseIdState[ 'baseId' ];
+type MappedInitialState = CompositeStoreProps & BaseIdState;
+
+export type CompositeState = CompositeStoreState & BaseIdState;
+export type CompositeStateProps =
+	| { state: CompositeState }
+	| ( CompositeState & { state?: never } );
+export type CompositeProps< C extends Component > = CompositeStateProps &
+	Parameters< C >[ 0 ];
```

Comment:
Nit: A bit more idiomatic this way?

```suggestion
	React.ComponentProps< C >;
```